server/src/main/java/org/elasticsearch/search/lookup/DocLookup.java

```diff
@@ -21,29 +21,28 @@
 import org.apache.lucene.index.LeafReaderContext;
 import org.elasticsearch.index.fielddata.IndexFieldData;
 import org.elasticsearch.index.mapper.MappedFieldType;
-import org.elasticsearch.index.mapper.MapperService;
 
 import java.util.function.Function;
 
-public class DocLookup {
+public final class DocLookup {
 
-    private final MapperService mapperService;
+    private final Function<String, MappedFieldType> fieldTypeLookup;
     private final Function<MappedFieldType, IndexFieldData<?>> fieldDataLookup;
 
-    DocLookup(MapperService mapperService, Function<MappedFieldType, IndexFieldData<?>> fieldDataLookup) {
-        this.mapperService = mapperService;
+    DocLookup(Function<String, MappedFieldType> fieldTypeLookup, Function<MappedFieldType, IndexFieldData<?>> fieldDataLookup) {
```

Comment:
I wonder if we could go even further in a follow up, and just have one String->IndexFieldData lookup?

Reply:
interesting idea, and this could be applied to other places too, where all we do is look up the type by name and then look up fielddata by field type. I will look into this.

Reply:
I looked into this and I found logic around the field type in the callers, so I will leave this for another time :)